File: Exercism Codes/C#/SgfParsing.cs
using System.Text.RegularExpressions;

public class SgfTree
{
    public SgfTree(IDictionary<string, string[]> data, params SgfTree[] children)
    {
        Data = data;
        Children = children;
    }

    public IDictionary<string, string[]> Data { get; }
    public SgfTree[] Children { get; set; }

    public override string ToString()
    {
        return $"Data: {string.Join(", ", Data.Select(x => $"{x.Key}=[{string.Join(", ", x.Value)}]"))}, " +
               $"Children: {Children.Length}";
    }
}

public class Property
{
    public string Label { get; set; }
    public string Kind { get; set; }

    public Property(string label, string kind)
    {
        Label = label;
        Kind = kind;
    }
}

public class SgfParser
{
    public static SgfTree ParseTree(string input)
    {
        if (input.Length == 0 || input[0] != '(' || input[^1] != ')')
        {
            throw new ArgumentException("tree missing");
        }
        SgfTree? sgf = ConstructNode(null, input.Replace("\\[", "[").Replace("\\]", "]")[1..^1]);
        if (sgf == null)
        {
            throw new ArgumentException("tree with no nodes");
        }
        return sgf;
    }

    private static SgfTree? ConstructNode(SgfTree? node, string input)
    {
        string? matcher = GetNode(input);
        List<string> parenthesis = ParseParenthesis(input);
        if (!string.IsNullOrEmpty(matcher))
        {
            node = ParseNode(matcher);
            SgfTree? childNode = ConstructNode(node, input.Replace(matcher, ""));
            if (childNode != null)
            {
                node.Children = node.Children.Concat([childNode]).ToArray();
            }
            return node;
        }
        if (parenthesis.Count > 0)
        {
            foreach (string par in parenthesis)
            {
                string extractPar = par[1..^1];
                SgfTree? childNode = ConstructNode(ParseNode(extractPar), extractPar);
                if (childNode != null)
                {
                    node.Children = node.Children.Concat([childNode]).ToArray();
                }
            }
        }
        return null;
    }

    private static string? GetNode(string input)
    {
        if (input.Length > 0 && input[0] != ';')
        {
            return null;
        }

        for (int i = 0; i < input.Length; i++)
        {
            if (i > 0 && i < input.Length && input.Substring(i - 1, 2) == "(;")
            {
                return input.Substring(0, i - 1);
            }
            if (i > 0 && i < input.Length && input.Substring(i - 1, 2) == "];")
            {
                return input.Substring(0, i);
            }
        }
        return input;
    }

    private static List<string> ParseParenthesis(string input)
    {
        List<string> nodes = new List<string>();

        while (!string.IsNullOrEmpty(input))
        {
            if (!input.StartsWith("(;"))
            {
                break;
            }

            int counter = 1;

            for (int i = 2; i < input.Length; i++)
            {
                if (input.Substring(i).StartsWith("(;"))
                {
                    counter++;
                }
                else if (input[i] == ')')
                {
                    counter--;
                }

                if (counter == 0)
                {
                    nodes.Add(input.Substring(0, i + 1));
                    input = input.Substring(i + 1);
                    break;
                }
            }
        }
        return nodes;
    }

    private static SgfTree ParseNode(string input)
    {
        IDictionary<string, List<string>> properties = new Dictionary<string, List<string>>();
        List<Property> matcher = GetMatches(input.Length >= 1 ? input[1..] : string.Empty);
        string? key = null;
        foreach (Property match in matcher)
        {
            if (match.Kind == "key")
            {
                if (match.Label.ToUpper() == match.Label)
                {
                    key = match.Label;
                }
                else
                {
                    throw new ArgumentException("property must be in uppercase");
                }
            }
            if (match.Kind == "value")
            {
                if (properties.ContainsKey(key))
                {
                    properties[key].Add(MapPropertyValue(match.Label));
                }
                else
                {
                    properties[key] = new List<string> { MapPropertyValue(match.Label) };
                }
            }
        }

        if (key != null && !properties.ContainsKey(key))
        {
            throw new ArgumentException("properties without delimiter");
        }

        return new SgfTree(properties.ToDictionary(x => x.Key, x => x.Value.ToArray()));
    }

    private static string MapPropertyValue(string input)
    {
        List<char> text = input[1..^1].ToCharArray().ToList();
        int i = 0;

        while (i < text.Count)
        {
            if (i > 0 && (text[i] == '\\' || text[i] == 't' || text[i] == 'n') && text[i - 1] == '\\')
            {
                text.RemoveAt(i - 1);
                continue;
            }
            if (i > 0 && text[i] == '\t' && text[i - 1] == '\\')
            {
                text.RemoveAt(i - 1);
                text.RemoveAt(i - 1);
                text.Insert(i - 1, ' ');
                continue;
            }
            if (i > 0 && text[i] == '\n' && text[i - 1] == '\\')
            {
                text.RemoveAt(i - 1);
                text.RemoveAt(i - 1);
                continue;
            }
            i++;
        }
        return Regex.Replace(string.Join("", text), "\\t", " ");
    }

    private static List<Property> GetMatches(string input)
    {
        List<Property> result = new List<Property>();
        int start = 0;

        for (int i = 0; i < input.Length; i++)
        {
            if (char.IsLetter(input[i]))
            {
                if (i == 0)
                {
                    start = i;
                    continue;
                }
                if (i > 0 && input[i - 1] == ']' && input[start] == '[')
                {
                    if (!HasClosingBeforeOpening(input, i))
                    {
                        result.Add(new Property(input.Substring(start, i - start), "value"));
                        start = i;
                        continue;
                    }
                }
            }
            if (input[i] == '[')
            {
                if (i > 0 && input[i - 1] == ']' && input[start] == '[')
                {
                    result.Add(new Property(input.Substring(start, i - start), "value"));
                    start = i;
                    continue;
                }
                if (i > 0 && char.IsLetter(input[i - 1]) && char.IsLetter(input[start]))
                {
                    result.Add(new Property(input.Substring(start, i - start), "key"));
                    start = i;
                    continue;
                }
            }
        }

        if (input.Length > 0 && input[^1] == ']')
        {
            result.Add(new Property(input[start..], "value"));
        }
        if (input.Length >= 1 && input.Length <= 2 && input.All(char.IsLetter))
        {
            result.Add(new Property(input, "key"));
        }

        return result;
    }

    private static bool HasClosingBeforeOpening(string input, int index)
    {
        for (int i = index; i < input.Length; i++)
        {
            if (input[i] == ']')
            {
                return true;
            }
            if (input[i] == '[')
            {
                return false;
            }
        }

        return false;
    }
}

File: Exercism Codes/Python/sgf_parsing.py
import re
import string


class SgfTree:
    def __init__(self, properties=None, children=None):
        self.properties = properties or {}
        self.children = children or []

    def __repr__(self, indent=None):
        return f"({self.properties}: {self.children})"

    def __eq__(self, other):
        if not isinstance(other, SgfTree):
            return False
        for key, value in self.properties.items():
            if key not in other.properties:
                return False
            if other.properties[key] != value:
                return False
        for key in other.properties.keys():
            if key not in self.properties:
                return False
        if len(self.children) != len(other.children):
            return False
        for child, other_child in zip(self.children, other.children):
            if child != other_child:
                return False
        return True

    def __ne__(self, other):
        return not self == other

class Property:
    def __init__(self, kind, label):
        self.kind = kind
        self.label = label


def parse(input_string: str) -> SgfTree:
    if not input_string or not input_string[0] == "(" or not input_string[-1] == ")":
        raise ValueError("tree missing")
    sgf = construct_node(None, input_string.replace("\\[", "[").replace("\\]", "]")[1:-1])
    if not sgf:
        raise ValueError("tree with no nodes")
    return sgf

def construct_node(node: SgfTree, input_string: str) -> SgfTree:
    matcher = get_node(input_string)
    parenthesis = parse_parenthesis(input_string)
    if matcher:
        node = parse_node(matcher)
        child_node = construct_node(node, input_string.replace(matcher, ""))
        if child_node:
            node.children.append(child_node)
        return node
    if parenthesis:
        for par in parenthesis:
            extract_par = par[1:-1]
            child_node = construct_node(parse_node(extract_par), extract_par)
            if child_node:
                node.children.append(child_node)
    return None

def get_node(text):
    if len(text) > 0 and text[0] != ";":
        return None

    for i in range(len(text)):
        if 0 < i < len(text) and text[i - 1:i + 1] == "(;":
            return text[:i - 1]
        if 0 < i < len(text) and text[i - 1:i + 1] == "];":
            return text[:i]

    return text


def parse_parenthesis(input_string: str) -> list[str]:
    nodes = []

    while input_string:
        if not input_string.startswith("(;"):
            break

        counter = 1

        for i in range(2, len(input_string)):
            if input_string.startswith("(;", i):
                counter += 1
            elif input_string[i] == ")":
                counter -= 1

            if counter == 0:
                nodes.append(input_string[:i + 1])
                input_string = input_string[i + 1:]
                break
        else:
            break

    return nodes

def parse_node(input_string: str) -> SgfTree:
    properties = {}
    matcher = get_matches(input_string[1:])
    key = None
    for match in matcher:
        if match.kind == "key":
            if match.label.isupper():
                key = match.label
            else:
                raise ValueError("property must be in uppercase")
        if match.kind == "value":
            if key in properties:
                properties[key].append(map_property_value(match.label))
            else:
                properties[key] = [map_property_value(match.label)]

    if key and key not in properties:
        raise ValueError("properties without delimiter")

    return SgfTree(properties)

def map_property_value(c: str) -> str:
    text = list(c[1:-1])
    i = 0

    while i < len(text):
        if i > 0 and (text[i] == "\\" or text[i] == "t" or text[i] == "n") and text[i - 1] == "\\":
            text.pop(i - 1)
            continue
        if i > 0 and text[i] == "\t" and text[i - 1] == "\\":
            text.pop(i - 1)
            text.pop(i - 1)
            text.insert(i - 1, " ")
            continue
        if i > 0 and text[i] == "\n" and text[i - 1] == "\\":
            text.pop(i - 1)
            text.pop(i - 1)
            continue
        i += 1
    return re.sub("\\t", " ", "".join(text))

def get_matches(text):
    result: list[Property] = []
    start = None

    for i, char in enumerate(text):
        if char in string.ascii_letters:
            if i == 0:
                start = i
                continue

            if i > 0 and text[i - 1] == ']' and text[start] == '[':
                if not has_closing_before_opening(text, i):
                    result.append(Property("value", text[start:i]))
                    start = i
                    continue

        if char == '[':
            if i > 0 and text[i - 1] == ']' and text[start] == '[':
                result.append(Property("value", text[start:i]))
                start = i
                continue

            if i > 0 and text[i - 1] in string.ascii_letters and text[start] in string.ascii_letters:
                result.append(Property("key", text[start:i]))
                start = i
                continue

    if len(text) > 0 and text[-1] == ']':
        result.append(Property("value", text[start:]))
    if 1 <= len(text) <= 2 and text.isalpha():
        result.append(Property("key", text))

    return result

def has_closing_before_opening(text, i):
    for x in range(i, len(text)):
        if text[x] == ']':
            return True
        if text[x] == '[':
            return False
    return False